rclcpp/include/rclcpp/experimental/intra_process_manager.hpp

```diff
@@ -344,12 +352,162 @@ class IntraProcessManager
 
   RCLCPP_PUBLIC
   void
-  insert_sub_id_for_pub(uint64_t sub_id, uint64_t pub_id, bool use_take_shared_method);
+  insert_sub_id_for_pub(
+    uint64_t sub_id, uint64_t pub_id, bool use_take_shared_method,
+    bool is_serialized_subscriber);
 
   RCLCPP_PUBLIC
   bool
   can_communicate(PublisherInfo pub_info, SubscriptionInfo sub_info) const;
 
+  template<
+    typename MessageT,
+    typename Alloc = std::allocator<void>,
+    typename Deleter = std::default_delete<MessageT>>
+  void
+  do_intra_process_publish_same_type(
+    std::unique_ptr<MessageT, Deleter> message,
+    std::shared_ptr<typename allocator::AllocRebind<MessageT, Alloc>::allocator_type> allocator,
+    const std::vector<uint64_t> & take_ownership_subscriptions,
+    const std::vector<uint64_t> & take_shared_subscriptions)
+  {
+    // subsriber and publisher have same message types
+    using MessageAllocTraits = allocator::AllocRebind<MessageT, Alloc>;
+    using MessageAllocatorT = typename MessageAllocTraits::allocator_type;
+
+    if (take_ownership_subscriptions.empty()) {
+      // None of the buffers require ownership, so we promote the pointer
+      std::shared_ptr<MessageT> msg = std::move(message);
+
+      this->template add_shared_msg_to_buffers<MessageT>(msg, take_shared_subscriptions);
+    } else if (!take_ownership_subscriptions.empty() && // NOLINT
+      take_shared_subscriptions.size() <= 1)
+    {
+      // There is at maximum 1 buffer that does not require ownership.
+      // So this case is equivalent to all the buffers requiring ownership
+
+      // Merge the two vectors of ids into a unique one
+      std::vector<uint64_t> concatenated_vector(take_shared_subscriptions);
```

Comment:
consider calling `reserve` here as the total amount of elements in the merged vector is known.would it also make sense to invert this here? Copying the `take_ownership_subscriptions` assuming that it's the bigger vector of the two.